### PR TITLE
Jetstream API

### LINKED ISSUE #352: Jetstream APIs
The Java API should be extended for Jetstream functionality.  The full feature set will require header implementation, #338.
 
- [ ] Implement the equivalent of the following golang APIs
  - [ ]  nats-io/nats.go#585
  - [ ]  nats-io/nats.go#596
  - [ ]  nats-io/nats.go#599
- [ ] Enhance CI to use master or nightly build of nats-server (required to test).
- [ ] Unit Tests

Rather than varargs like golang, I'd suggest overloading the existing publish and subscribe APIs.

Some suggestions:

New APIs (add appropriate overloads for timeouts, contexts)
### Publish
- [ ] Connection.Publish - add overloads for timeout and context (or builder APIs?)  Note that overloads are likely easiest for the user and will perform better.

### Message
- [ ] Message.getJetStreamMetaData()
- [ ] Message.Ack
- [ ] Message.Nak
- [ ] Message.AckProgress
- [ ] Message.AckAndFetch
- [ ] Message.AckTerm
- [ ] static Message.ParsePublishAck (or static Connection.ParsePublishAck or Nats.ParsePubishAck())

### Consumer
- [ ] Subscriber.CreateConsumerConfiguration()
- [ ] ConsumerConfiguration.setDurable(string)
- [ ] ConsumerConfiguration.setDeliveryPolicy(DeliveryPolicy)
- [ ] ConsumerConfiguration.setStartSeq(int)
- [ ] ConsumerConfiguration.setStartTime(TBD - see STAN)
- [ ] ConsumerConfiguration.setAckPolicy(AckPolicy)
- [ ] ConsumerConfiguration.setAckWait(Duration)
- [ ] ConsumerConfiguration.setMaxDelivery(int)
- [ ] ConsumerConfiguration.setFilterSubject(string)
- [ ] ConsumerConfiguration.setReplayPolicy(ReplayPolicy)
- [ ] ConsumerConfiguration.SetSampleFrequency(string)
- [ ] ConsumerConfiguration.SetRateLimit(long)

Overload the subscribe APIs with a stream and consumer configuration
- [ ] Subscribe(..., string stream, ConsumerConfiguration)

### PR DESCRIPTION
**Pull request by ColinSullivan1**
## Feature:  The Jetstream API

Documentation will be incomplete until the API is finalized.

Resolves #352 

### TODO
- [x] Consumer Configuration
- [x] ProducerOptions
- [x] SubscriberOptions
- [x] Ack
- [x] Publish APIs
- [x] Test Harness Updates
- [x] Msg APIs
- [x] Subscribe APIs
- [x] Pull Consumer

### CODE REVIEW

**Comment by jameshilliard** [comment]
Looks like some of the tests failures are caused by the issue fixed in #348.

**Comment by ColinSullivan1** [comment]
Thanks, I'll add that locally here for now...

File: src/main/java/io/nats/client/impl/Ack.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    private static final String grabString = "\\s*\"(.+?)\"";
+    private static final String grabNumber = "\\s*(\\d+)";
+
+    private static final Pattern streamRE = Pattern.compile("\"stream\":" + grabString, Pattern.CASE_INSENSITIVE);
+    private static final Pattern seqnoRE = Pattern.compile("\"seq\":" + grabNumber, Pattern.CASE_INSENSITIVE); 
+
+    // Acks will be received with the following format:
+    // "-ERR <server message>""
+    // "+OK { "stream" : "mystream", "seq", 42}"
```
[the remark this answers is not in the document]
Comment (ColinSullivan1) [reply src/main/java/io/nats/client/impl/Ack.java:34]: Will fix...

File: src/main/java/io/nats/client/PublishOptions.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+         * @param stream The name of the stream.
+         * @return Builder
+         */
+        public Builder stream(String stream) {
+            this.stream = stream;
+            return this;
+        }
+
+        /**
+         * Sets the timeout to wait for a publish acknoledgement from a jetstream
```
[the remark this answers is not in the document]
Comment (ColinSullivan1) [reply src/main/java/io/nats/client/PublishOptions.java:129]: thx..

File: src/main/java/io/nats/client/ConsumerConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+    private String deliverSubject = null;
+	private long startSeq = 0;
+	private ZonedDateTime startTime = null;
+	private AckPolicy ackPolicy = AckPolicy.Explicit;
+	private Duration ackWait = Duration.ofSeconds(30);
+	private long maxDeliver = 0;
+	private String filterSubject = null;
+	private ReplayPolicy replayPolicy = ReplayPolicy.Instant;
+	private String sampleFrequency = null;
+    private long rateLimit = 0;
```
Comment (scottf) [inline src/main/java/io/nats/client/ConsumerConfiguration.java:146]: nit: indentation

File: src/main/java/io/nats/client/ConsumerConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,644 @@
...
+            cc.replayPolicy, cc.sampleFrequency, cc.rateLimit, cc.deliverSubject,
+            cc.maxWaiting, cc.maxAckPending);
+    }
+
+    // For the builder
+    ConsumerConfiguration(String durable, DeliverPolicy deliverPolicy, long startSeq,
+            ZonedDateTime startTime, AckPolicy ackPolicy, Duration ackWait, long maxDeliver, String filterSubject,
+            ReplayPolicy replayPolicy, String sampleFrequency, long rateLimit, String deliverSubject,
+            long maxWaiting, long maxAckPending) {
+
```
Comment (scottf) [inline src/main/java/io/nats/client/ConsumerConfiguration.java:272]: lots of constructor parameters is a good use case for builders

File: src/main/java/io/nats/client/Options.java [hunk 5]
```diff
@@ -1773,4 +1773,12 @@ private void appendOption(CharBuffer builder, String key, char[] value, boolean
...
         }
     }
+
+    /**
+     * Set old request style.
+     * @param value true to use the old request style
+     */
+    public void setOldRequestStyle(boolean value) {
+        useOldRequestStyle = value;
+    }
```
Comment (scottf) [inline src/main/java/io/nats/client/Options.java:1783]: do you want to allow change style once in use? useOldRequestStyle defaults to false and the builder sets it to true.

File: src/main/java/io/nats/client/Message.java [hunk 6]
```diff
@@ -26,6 +29,56 @@
  */
 public interface Message {
 
+	/**
+	 * This interface defined the methods use to get Jetstream metadata about a message, 
+	 * when applicable.
+	 */
+	public interface MetaData {
```
Comment (scottf) [inline src/main/java/io/nats/client/Message.java:36]: I would pull this out into it's own file for sure.

File: src/main/java/io/nats/client/Nats.java [hunk 7]
```diff
@@ -14,6 +14,7 @@
 package io.nats.client;
 
 import java.io.IOException;
+import java.time.format.DateTimeFormatter;
```
Comment (scottf) [inline src/main/java/io/nats/client/Nats.java:17]: I think this is unused and can be removed.

File: src/test/java/io/nats/client/SubscriberTests.java [hunk 8]
```diff
@@ -419,7 +419,7 @@ public void testThrowOnNullQueue() {
         assertThrows(IllegalArgumentException.class, () -> {
             try (NatsTestServer ts = new NatsTestServer(false);
                         Connection nc = Nats.connect(ts.getURI())) {
-                nc.subscribe("subject", null);
+                nc.subscribe("subject", (String)null);
```
Comment (scottf) [inline src/test/java/io/nats/client/SubscriberTests.java:422]: I don't think this cast is needed

**Review by scottf — APPROVED** [review]
LGTM

File: src/main/java/io/nats/client/ConsumerInfo.java [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+/**
+ * The ConsumerInfo class returns information about a jetstream consumer.
+ */
+public class ConsumerInfo {
+
+    /**
+     * This class holds the sequence numbers for a consumer and 
+     * stream.
+     */
+    public class SequencePair {
```
Comment (scottf) [inline src/main/java/io/nats/client/ConsumerInfo.java:36]: This is sort nitty.I don't think this needs to be public, and it's not strictly an inner class in it's usage, more like a companion.
```
private static class SequencePair {
```

File: src/main/java/io/nats/client/impl/NatsMessage.java [hunk 10]
```diff
@@ -260,6 +289,224 @@ public Subscription getSubscription() {
...
+    public void inProgress() {
+        ackReply(AckProgress);
+    }
+
+    @Override
+    public void term() {
+        ackReply(AckTerm);
+    }
+
+    public void toDOackNextRequest(ZonedDateTime expiry, long batch, boolean noWait) {
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/NatsMessage.java:380]: is this function name correct?

File: src/main/java/io/nats/client/impl/NatsMessage.java [hunk 11]
```diff
@@ -260,6 +289,224 @@ public Subscription getSubscription() {
...
+        if (c == null) {
+            throw new IllegalStateException("Message is not bound to a connection");
+        }
+        
+        // minor optimization for the ack.
+        byte[] payload;
+        if (expiry == null && batch == 0 && !noWait) {
+            payload = AckNextEmptyPayload;
+        } else {
+            StringBuilder sb = new StringBuilder("+ACKNXT {");
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/NatsMessage.java:396]: good candidate for ByteArrayBuilder

Comment (ColinSullivan1) [reply src/main/java/io/nats/client/impl/NatsMessage.java]: Yeah...  I ended up refactoring this code entirely so there's no string builder anymore here.

File: src/main/java/io/nats/client/Options.java (hunk 5 above)
Comment (ColinSullivan1) [reply src/main/java/io/nats/client/Options.java:1783]: I have to for Jetstream.  Right now it always needs to be enabled to allow subject overrides, which JS does.

**Comment by ripienaar** [comment]
LGTM

**Pull request merged**